content/blog/callpackage-a-tool-for-the-lazy.md

```diff
@@ -0,0 +1,219 @@
+---
+title: callPackage, a tool for the lazy
+date: 2022-08-25
+extra:
+  author: Norbert Melzer
+---
+
+In [`nixpkgs`](https://github.com/nixos/nixpkgs), there is a massive use of the
+`callPackage` function, which provides us with a lot of benefits.
+
+## Basic examples
+
+Though before even discussing the benefits, lets see how it actually gets used.
+
+Given are these 2 files `hello.nix` and `default.nix`:
+
+```nix
+# Usually I would prefer the usage of nix flakes, though that would
+# introduce a lot of boilerplate in the examples for no reason,
+# therefore I will use diamond paths and `import` for simplicity.
+```
+
+```nix
+# default.nix
+let pkgs = import <nixpkgs> {}; in
+pkgs.callPackage ./hello.nix {}
+```
+
+```nix
+# hello.nix
+{writeShellScriptBin}:
+writeShellScriptBin "hello" ''
+  echo "hello, world!"
+''
+```
+
+Building using `nix-build` will produce `./result/bin/hello`, and running it will
+nicely greet you.
+
+As you can see, `writeShellScriptBin` gets passed in by `callPackage` automatically.
```

Comment:
On a personal note, this is really auto*magical* and potentially very very confusing if you don't know that. Great that you decided to write this article.

Reply:
I will sit down during lunch time and see if I can rephrase this paragraph to explain a tad more without making the post explode.

I explicitely wanted to avoid "how does it work" in favor of "why do you want to use it"

Reply:
> I explicitely wanted to avoid "how does it work" in favor of "why do you want to use it"

That's fine, but I think a bird's eye view is in order. I think I added something in one of the suggestions already, you may pull that out here.

Reply:
I am not a friend of the term "automagically", ruby spoiled me.

I hope 91be10f64694cb33fdf024e382f6106978435d10 is still in the spirit of your suggested changes.